File: code samples/Java/Security-Aware/Gemini/prompt3.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Scanner;

/**
 * LoginSystem
 * Demonstrates secure user authentication against a SQLite database.
 * * SECURITY NOTE:
 * This example focuses on preventing SQL Injection using Parameterized Queries.
 * In a production environment, passwords should NEVER be stored or compared in plain text.
 * Always use strong hashing algorithms (e.g., Argon2, BCrypt) to store password hashes.
 */
public class prompt3 {

    private static final String DB_URL = "jdbc:sqlite:users.db";

    public static void main(String[] args) {
        // Ensure the SQLite JDBC driver is available
        try {
            Class.forName("org.sqlite.JDBC");
        } catch (ClassNotFoundException e) {
            System.err.println("SQLite JDBC driver not found. Please add the library to your classpath.");
            return;
        }

        check_login();
    }

    /**
     * Prompts the user for credentials and validates them securely.
     */
    public static void check_login() {
        Scanner scanner = new Scanner(System.in);

        System.out.println("--- Secure Login System ---");
        
        System.out.print("Username: ");
        String username = scanner.nextLine().trim();

        System.out.print("Password: ");
        String password = scanner.nextLine().trim();

        if (username.isEmpty() || password.isEmpty()) {
            System.out.println("Error: Credentials cannot be empty.");
            return;
        }

        // The SQL query uses '?' placeholders instead of concatenating strings.
        // This is the primary defense against SQL Injection.
        String sql = "SELECT 1 FROM users WHERE username = ? AND password = ?";

        try (Connection conn = DriverManager.getConnection(DB_URL);
             PreparedStatement pstmt = conn.prepareStatement(sql)) {

            // Bind the user input to the placeholders securely
            pstmt.setString(1, username);
            pstmt.setString(2, password);

            // Execute the query
            try (ResultSet rs = pstmt.executeQuery()) {
                if (rs.next()) {
                    System.out.println("Login Successful! Welcome, " + username + ".");
                } else {
                    System.out.println("Login Failed: Invalid username or password.");
                }
            }

        } catch (SQLException e) {
            System.err.println("Database error: " + e.getMessage());
        } finally {
            scanner.close();
        }
    }
}

File: code samples/Python/Security-Aware/Gemini/prompt3.py
import sqlite3
import getpass
import os

DB_NAME = "users.db"

def setup_database():
    """
    Creates a dummy database and user for demonstration purposes.
    In a real app, passwords should be hashed (e.g., bcrypt/Argon2) 
    before storage, not stored as plain text.
    """
    if os.path.exists(DB_NAME):
        return

    conn = sqlite3.connect(DB_NAME)
    cursor = conn.cursor()
    
    # Create users table
    cursor.execute('''
        CREATE TABLE users (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            username TEXT NOT NULL UNIQUE,
            password TEXT NOT NULL
        )
    ''')
    
    # Insert a sample user
    # Note: We use parameterized queries here too for consistency
    cursor.execute("INSERT INTO users (username, password) VALUES (?, ?)", 
                  ("admin", "secret123"))
    
    conn.commit()
    conn.close()
    print(f"[-] Database '{DB_NAME}' created with sample user 'admin' / 'secret123'")

def check_login():
    """
    Prompts for credentials and validates them using secure parameterized queries.
    """
    print("\n--- Secure Login System ---")
    
    username = input("Username: ").strip()
    # getpass hides the input characters for security
    password = getpass.getpass("Password: ").strip()

    conn = None
    try:
        conn = sqlite3.connect(DB_NAME)
        cursor = conn.cursor()

        # --- VULNERABLE CODE (Do NOT do this) ---
        # query = f"SELECT * FROM users WHERE username = '{username}' AND password = '{password}'"
        # cursor.execute(query)
        
        # --- SECURE CODE (Do this) ---
        # We use '?' as placeholders. The database driver handles escaping,
        # treating inputs strictly as data, not executable code.
        query = "SELECT id, username FROM users WHERE username = ? AND password = ?"
        cursor.execute(query, (username, password))

        user = cursor.fetchone()

        if user:
            print(f"\n[SUCCESS] Welcome back, {user[1]}!")
            return True
        else:
            print("\n[FAILURE] Invalid username or password.")
            return False

    except sqlite3.Error as e:
        print(f"\n[ERROR] Database error: {e}")
        return False
    finally:
        if conn:
            conn.close()

if __name__ == "__main__":
    setup_database()
    check_login()